tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -607,6 +631,39 @@ Status StorageManager::touch(const URI& uri) {
   return vfs_->touch(uri);
 }
 
+Status StorageManager::check_array_encryption_key(
+    const ArraySchema* schema, const void* encryption_key, bool was_cache_hit) {
+  if (schema->encryption_type() == EncryptionType::NO_ENCRYPTION &&
+      encryption_key != nullptr)
+    return LOG_STATUS(Status::StorageManagerError(
+        "Encryption key check failed; array not encrypted"));
+
+  std::string uri = schema->array_uri().to_string();
+  EncryptionKeyValidation* validation = nullptr;
+
+  // Get or add the validation instance
+  std::lock_guard<std::mutex> lck(open_arrays_encryption_keys_mtx_);
+  auto it = open_arrays_encryption_keys_.find(uri);
+  if (it == open_arrays_encryption_keys_.end()) {
+    // Sanity check for cached schemas, which should already have added a
+    // validation instance.
+    if (was_cache_hit)
+      return LOG_STATUS(
+          Status::StorageManagerError("Encryption key check failed; schema was "
+                                      "cached but key not previously used."));
+
+    auto ptr = std::unique_ptr<EncryptionKeyValidation>(
+        new EncryptionKeyValidation(schema->encryption_type()));
+    validation = ptr.get();
+    open_arrays_encryption_keys_[uri] = std::move(ptr);
```

Comment:
Hmm, this seems to be growing indefinitely as new encrypted arrays are being opened. I think we should remove the entry when an array is closed.

Reply:
Hmm, I think this is more complicated than it seems. Suppose an encrypted array is opened properly (schema goes into cache) and then closed. If we remove this "validation" information when the array is closed, a subsequent open with the *wrong* key would get the schema from the cache, and then compute *new* "validation" information using the wrong key (which would succeed, as there's nothing to compare against).

Reply:
I see, then this should be invalidated when the array schema is evicted from the cache.